tests/test_generate_data.py

```diff
@@ -3,12 +3,464 @@
+
+
+def validate_recipe(recipe_file_name):
+    with open(recipe_file_name, encoding="utf-8") as fp:
+        yaml_contents = yaml.safe_load(fp)
+        assert len(yaml_contents["datasets"]) == 1
+        assert yaml_contents["datasets"][0]["path"].endswith(".jsonl")
+        assert "sampling_size" in yaml_contents["datasets"][0]
+        assert yaml_contents["metadata"]["sys_prompt"] == _SYS_PROMPT
+
+
+def validate_mixed_dataset(dataset_file_name):
+    ds = load_dataset("json", data_files=dataset_file_name, split="train")
+    assert "messages" in ds.features
+    assert len(ds.features["messages"]) == 1
+    assert len(ds.features["messages"][0]) == 2
+    assert ds.features["messages"][0]["content"].dtype == "string"
+    assert ds.features["messages"][0]["role"].dtype == "string"
+
+
+def validate_lm_eval_task(lm_eval_task_file_name):
+    with open(lm_eval_task_file_name, encoding="utf-8") as fp:
+        yaml_contents = yaml.safe_load(fp)
+        assert "task" in yaml_contents
+        assert "dataset_kwargs" in yaml_contents
+        assert "doc_to_text" in yaml_contents
+        assert "doc_to_choice" in yaml_contents
+        assert "doc_to_target" in yaml_contents
+
+
+def validate_mmlubench_dataset(dataset_file_name):
+    with open(dataset_file_name, encoding="utf-8") as fp:
+        # FIXME: fix the mmlubench pipeline in this test
+        assert fp.readlines() == []
+
+
+def generate_test_samples(yaml_contents):
+    """Convert questions and answers from the taxonomy format into the
+    user/assistant format used by the legacy training methods such as
+    Linux training and MacOS training.
+
+    This mirrors what _gen_test_data() does.
+    """
+    test_samples = []
+    is_knowledge = "document" in yaml_contents
+    for seed_example in yaml_contents["seed_examples"]:
+        if is_knowledge:
+            for qna in seed_example["questions_and_answers"]:
+                test_samples.append(
+                    {
+                        "user": qna["question"]
+                        + "\n"
+                        + seed_example["context"].strip(),
+                        "assistant": qna["answer"].strip(),
+                    }
+                )
+
+        else:
+            # FIXME: handle freeform skills - no context
+            test_samples.append(
+                {
+                    "user": seed_example["question"] + "\n" + seed_example["context"],
+                    "assistant": seed_example["answer"],
+                }
+            )
+    return test_samples
+
+
+def generate_train_samples(yaml_contents):
+    """Generate expected training samples in the user/assistant format
+    used by the legacy training methods such as Linux training and MacOS
+    training.
+
+    Mirroring _noop_llmblock_generate() below, we generate 10 samples
+    per input, and then follow _gen_train_data()'s output format.
+    """
+
+    def add_q(q):
+        return (q + "?") if not "?" in q else q
```

Comment:
nit:
```suggestion
    def add_qestion_mark(q):
        if not q.endswith("?"):
            return q + "?"
        return q
```

Reply:
just want to add I assumed you were trying to specifically match strings that don't end with a question mark. If you instead mean to search for a question mark anywhere within the string, you can disregard this comment. Although I still think it's slightly more readable as a multiline `if`.

Reply:
Its a valid point that you make. I think however the  use of the ternary operator in this case is simple enough to be understandable.  With that said, I have updated the inner function name from `add_q` to `add_question_mark` for better clarity.